### PR TITLE
Monitor pipelines

### PR DESCRIPTION
**Pull request by JonathanGriffe**

### CODE REVIEW

File: src/api.js [hunk 1]
```diff
@@ -17,33 +18,376 @@ const getMessages = args => {
...
+        actorName += (name.includes('|') ? '(' + name + ')' : name) + '[' + nameDict[name] + '] ';
+      }
+      if (Object.keys(nameDict).length > 1) {
+        actorName = '[' + actorName + ']';
+      }
+      composition.actorName = actorName;
+    } else {
+      let actorName = '';
+      for (const name of composition.messages.map(message => message.actorName)) {
+        actorName += name + ' | ';
```
Comment (SimonDosda) [inline src/api.js:139]: This should work no ?
```js
 composition.actorName = composition.messages.map(({actorName}) => actorName).join(' | ');
```

File: src/api.js [hunk 2]
```diff
@@ -17,33 +18,376 @@ const getMessages = args => {
...
+          message.startedDatetime ? message.startedDatetime.valueOf() : Infinity
+        )
+      );
+      if (datetime !== Infinity) {
+        composition.startedDatetime = new Date(datetime);
+      }
+    }
+    // add Wait time
+    let wait_time = 0;
+    composition.messages.forEach(message => {
```
Comment (SimonDosda) [inline src/api.js:182]: use `reduce`

File: src/api.js [hunk 3]
```diff
@@ -17,33 +18,376 @@ const getMessages = args => {
...
+      if (message.waitTime) {
+        wait_time += message.waitTime;
+      }
+    });
+    composition.waitTime = wait_time;
+    // add Execution time
+    let execution_time = 0;
+    composition.messages.forEach(message => {
```
Comment (SimonDosda) [inline src/api.js:190]: same

File: src/api.js [hunk 4]
```diff
@@ -17,33 +18,376 @@ const getMessages = args => {
...
+      if (message.executionTime) {
+        execution_time += message.executionTime;
+      }
+    });
+    composition.executionTime = execution_time;
+    //add Remaining Time
+    if (composition.messages.every(message => message.endDatetime || message.remainingTime)) {
+      let remaining_time = 0;
+      composition.messages.forEach(message => {
```
Comment (SimonDosda) [inline src/api.js:199]: same

File: src/api.js [hunk 5]
```diff
@@ -17,33 +18,376 @@ const getMessages = args => {
...
+    }
+    // add Progress
+    if (
+      composition.messages.every(message => message.status === 'Success' || message.progress) &&
+      !composition.messages.every(message => message.status === 'Success')
+    ) {
+      composition.progress =
+        composition.messages
+          .map(message => message.progress || 1)
+          .reduce((value, acc) => value + acc) / composition.messages.length;
```
Comment (SimonDosda) [inline src/api.js:214]: value and acc are inversed here. I would merge with the map 
```suggestion
          .reduce((acc, message) => ACC + (message.progress || 1)) / composition.messages.length;
```

File: src/api.js [hunk 6]
```diff
@@ -17,33 +18,376 @@ const getMessages = args => {
...
+    }
+  }
+
+  Object.entries(compositions).forEach(composition_messages => {
+    const composition = assembleComposition(0, composition_messages[1]);
+    messages.push(composition);
+  });
+
+  return { messages: messages, loadDateTime: loadDateTime };
+}
```
Comment (SimonDosda) [inline src/api.js:374]: ok this is quite long, I will read all this later. Put it in his own file and add a docstring to each module so we know what it is used for.

File: src/api.js [hunk 7]
```diff
@@ -121,6 +451,42 @@ const cleanStates = args => {
...
+    daily_time: job.dailyTime,
+    enabled: job.enabled,
+    interval: job.interval,
+    iso_weekday: job.isoWeekday,
+    kwargs: job.kwargs,
+    lastQueued: job.last_queued,
+    tz: job.tz
+  };
+  if (parsedJob.daily_time) {
+    parsedJob.interval = 86400;
```
Comment (SimonDosda) [inline src/api.js:467]: would null work ?

File: src/api.js [hunk 8]
```diff
@@ -17,33 +17,433 @@ const getMessages = args => {
   const url = '/messages/states';
   return axios
     .post(url, args)
-    .then(res => ({ count: res.data.count, data: res.data.data.map(parseMessage) }));
+    .then(res => Object.assign(parseMessages(res.data.data), { count: res.data.count }));
```
Comment (SimonDosda) [inline src/api.js:20]:
```suggestion
    .then(res => ({...parseMessages(res.data.data), count: res.data.count }));
```

Comment (SimonDosda) [reply src/api.js:20]: count should probably be returned by parseMessages

File: src/api.js [hunk 9]
```diff
@@ -17,33 +17,433 @@ const getMessages = args => {
...
+   * Recursively adds to the composition all the messages from the pipe target that have not been enqueued
+   * @param composition_id
+   * @param message
+   */
+  function fillPipe(message) {
+    if (message.pipeTarget) {
+      message.pipeTarget.forEach(pipe_element => {
+        if (
+          compositions[message.compositionId].findIndex(
+            composition_element => composition_element.messageId === pipe_element.messageId
```
Comment (SimonDosda) [inline src/api.js:130]: you can also use destructuring
```suggestion
            ({messageId}) => messageId === targetMessage.messageId
```

File: src/api.js [hunk 10]
```diff
@@ -17,33 +17,433 @@ const getMessages = args => {
...
+   * Assemble a composition
+   * @param composition_index
+   * @param composition_msgs
+   * @returns {*}
+   */
+  function assembleComposition(composition_index, composition_msgs) {
+    // finding the index of one of the first messages of a composition
+    let index = 0;
+    let prev_index = findPreviousElement(index, composition_msgs);
+    while (prev_index !== -1) {
```
Comment (SimonDosda) [inline src/api.js:382]: looks like a `do while`

File: src/api.js [hunk 11]
```diff
@@ -17,33 +17,433 @@ const getMessages = args => {
...
+    return addDetails(group);
+  }
+
+  /**
+   * Assemble a composition
+   * @param composition_index
+   * @param composition_msgs
+   * @returns {*}
+   */
+  function assembleComposition(composition_index, composition_msgs) {
```
Comment (SimonDosda) [inline src/api.js:378]: composition_index is not used (and carefull about camelCase)

File: src/api.js [hunk 12]
```diff
@@ -17,33 +17,433 @@ const getMessages = args => {
...
+    // finding the index of one of the first messages of a composition
+    let index = 0;
+    let prev_index = findPreviousElement(index, composition_msgs);
+    while (prev_index !== -1) {
+      index = prev_index;
+      prev_index = findPreviousElement(index, composition_msgs);
+    }
+    // If the composition is a group, find all the other messages in the group
+    if (composition_msgs[index].groupId) {
```
Comment (SimonDosda) [inline src/api.js:387]:
```suggestion
    const currentMessage = compositionMessages[index];
    if (currentMessage.groupId) {
```

File: src/api.js [hunk 13]
```diff
@@ -17,33 +17,433 @@ const getMessages = args => {
...
+      const group_indexes = [];
```
Comment (SimonDosda) [inline src/api.js:388]:
```suggestion
      const group_indexes = compositionMessages.reduce((indexes, message, index) => {
      if (message.groupId === currentMessage.groupId) {
          indexes.push(index);
          }
          return indexes
      }, []);
```

File: src/api.js [hunk 14]
```diff
@@ -17,33 +17,410 @@ const getMessages = args => {
...
+
+  /**
+   * Assemble the group starting with the given ids
+   * @param start_ids
+   * @param composition_msgs
+   * @returns {*}
+   */
+  function assembleGroup(start_ids, composition_msgs) {
+    const group = { type: 'group', messages: [] };
+    const groupId = findGroupId(start_ids, composition_msgs);
```
Comment (SimonDosda) [inline src/api.js:339]: I don't understand why this is so complicated. Same for the code after. Is it to manage group of groups?

File: src/api.js (hunk 8 above)
Comment (JonathanGriffe) [reply src/api.js:20]: This syntax is not available in es6

**Pull request merged**